### PR TITLE
[sglang] feat: add preparation for sglang+verl

### PR DESCRIPTION
**Pull request by lbk-sys**
### What does this PR do?
support npu for verl + sglang

```python
bash examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_npu.sh
```

### Accuracy test
8b：
8b

30b：
30b

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_npu.sh [hunk 1]
```diff
@@ -0,0 +1,57 @@
+set -x
+logs=/home/l00878165/sglang/repos/logs/jingdu0915/tmp1_qwen3_8b_512_1k_graph.log
```
Comment (FightingZhen) [inline examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_npu.sh]: Pay attention to private information protection, suggest using a specific path

File: examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_npu.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+export HCCL_CONNECT_TIMEOUT=1500
+
+num=4
```
Comment (FightingZhen) [inline examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_npu.sh]: This param is confusing, maybe define one or two params?

File: examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_npu.sh [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    trainer.experiment_name='qwen2_5_7b_function_rm' \
+    trainer.n_gpus_per_node=$num \
+    trainer.nnodes=1 \
+    trainer.save_freq=1000 \
+    trainer.test_freq=10000 \
+    trainer.total_epochs=5 \
+    trainer.default_local_dir="${CKPTS_DIR}" \
+    actor_rollout_ref.actor.ulysses_sequence_parallel_size=${sp_size} \
+    actor_rollout_ref.ref.ulysses_sequence_parallel_size=${sp_size} \
+    trainer.device=npu  2>&1 | tee -i $logs
```
Comment (FightingZhen) [inline examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_npu.sh]: Shell scripts usually not redirect console logs to specific file by default.

File: docs/ascend_tutorial/ascend_sglang_quick_start.rst [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+基础环境准备
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
++-----------+------------------+
+| software  | version          |
++-----------+------------------+
+| Python    | == 3.11          |
++-----------+------------------+
+| CANN      | == 8.3.RC1.B060  |
```
Comment (FightingZhen) [inline docs/ascend_tutorial/ascend_sglang_quick_start.rst]: CANN和驱动的B版本对社区其实应该是不感知的，这样写其实有点问题，看是否必须呈现，或者给出8.3.RC1的外部预计可用时间

File: docs/ascend_tutorial/ascend_sglang_quick_start.rst [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
++-----------+------------------+
+| HDK       | == 25.3.RC1.B061 |
++-----------+------------------+
+| torch     | == 2.6.0         |
```
Comment (FightingZhen) [inline docs/ascend_tutorial/ascend_sglang_quick_start.rst]: torch_npu的下一个长期维护版本是2.7.1，这里是必须使用2.6.0么？

File: docs/ascend_tutorial/ascend_sglang_quick_start.rst [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
++-----------+------------------+
+| torch_npu | == 2.6.0         |
++-----------+------------------+
+
+**目前verl框架中sglang npu后端仅支持上述HDK、CANN和PTA版本, 具体商发版本待发布**
+
+为了能够在 verl 中正常使用 sglang，需使用以下命令安装sglang、torch_memory_saver和verl。请注意根据机器类型区分安装方式。
```
Comment (FightingZhen) [inline docs/ascend_tutorial/ascend_sglang_quick_start.rst]: 下方的安装指引并未区分机器类型，此处描述建议删除

File: examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_spmd_npu.sh [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    actor_rollout_ref.actor.use_torch_compile=False \
+    actor_rollout_ref.model.enable_gradient_checkpointing=True \
+    actor_rollout_ref.actor.fsdp_config.param_offload=True \
+    actor_rollout_ref.actor.fsdp_config.optimizer_offload=True \
+    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=2 \
+    actor_rollout_ref.rollout.tensor_model_parallel_size=$tp_size \
+    actor_rollout_ref.rollout.name=sglang \
+    actor_rollout_ref.rollout.gpu_memory_utilization=0.3 \
+    actor_rollout_ref.rollout.n=5 \
+    +actor_rollout_ref.rollout.engine_kwargs.sglang.attention_backend="ascend" \
```
Comment (FightingZhen) [inline examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_spmd_npu.sh:51]: 这个参数现在应该已经是一个固定参数了，确认是否必须添加加号前缀

File: examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_npu.sh [hunk 8]
```diff
@@ -0,0 +1,67 @@
+set -x
+export HCCL_CONNECT_TIMEOUT=1500
+export HCCL_HOST_SOCKET_PORT_RANGE=60000-60050
+export HCCL_NPU_SOCKET_PORT_RANGE=61000-61050
+# export VERL_PPO_LOGGING_LEVEL=DEBUG
```
Comment (FightingZhen) [inline examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_npu.sh]: 不需要的环境变量建议删除

File: examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_npu.sh [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+model_path=$DATA_HOME/models/Qwen3-8B
+train_data=$DATA_HOME/datasets/processed_gsm8k/train.parquet
+valid_data=$DATA_HOME/datasets/processed_gsm8k/test.parquet
+
+python3 -m verl.trainer.main_ppo \
+    algorithm.adv_estimator=grpo \
+    data.train_files=$train_data \
+    data.val_files=$valid_data \
+    data.train_batch_size=16 \
+    data.max_prompt_length=512 \
```
Comment (FightingZhen) [inline examples/grpo_trainer/run_qwen3_8b_grpo_sglang_1k_npu.sh]: 1k和32k脚本的格式建议统一，在前面独立变量来承载prompt_length和response_length，其他的维度也检查对齐下

File: verl/utils/npu_utils.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    """
+    # dim = hidden_states.shape[-1]
+    # output = torch.zeros((batch * seqlen), dim, device=hidden_states.device, dtype=hidden_states.dtype)
+    # output[indices] = hidden_states
+    output = index_put_first_axis(hidden_states, indices, batch * seqlen)
+    return rearrange(output, "(b s) ... -> b s ...", b=batch)
+
+
+# Copied from https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/bert_padding.py
+def unpad_input(hidden_states, attention_mask, unused_mask=None):
```
Comment (wuxibin89) [inline verl/utils/npu_utils.py:99]: These functions already added in https://github.com/volcengine/verl/pull/3608

**Review by wuxibin89 — APPROVED** [review]
